src/BuiltInTools/dotnet-watch/HotReload/CompilationHandler.cs

```diff
@@ -213,12 +213,35 @@ private async ValueTask<bool> EnsureSolutionInitializedAsync()
 
         private async ValueTask<SourceText> GetSourceTextAsync(string filePath)
         {
+            var zeroLengthRetryPerformed = false;
             for (var attemptIndex = 0; attemptIndex < 6; attemptIndex++)
             {
                 try
                 {
-                    using var stream = File.OpenRead(filePath);
-                    return SourceText.From(stream, Encoding.UTF8);
+                    // File.OpenRead opens the file with FileShare.Read. This may prevent IDEs from saving file
+                    // contents to disk
+                    SourceText sourceText;
+                    using (var stream = File.Open(filePath, FileMode.Open, FileAccess.Read, FileShare.ReadWrite))
+                    {
+                        sourceText = SourceText.From(stream, Encoding.UTF8);
```

Comment:
Do we need to generate the `SourceText` from the stream or can we just check `stream.Length`?

Reply:
stream.Length can throw for non-seekable files. It seems much easier to rely on SourceText which always provides a value.